Service annotation based pod selection for NodePortLocal

shubhamavi opened the pull request:
Currently the NPLController in antrea-agent watches for all pods on the node, and exposes all pods to the external network using nodeip:nodeport.
This PR introduces pod selection/filtering for NodePortLocal based on an annotation to be provided in the Service objects. The Service objects can be of any type but `NodePort`. The proposed annotation NPLController will watch for is:
`nodeportlocal.antrea.io/enabled: "true"`

**Note**: The value "true" is case sensitive here
This makes the selection and implementation of NodePortLocal much more controlled.

tnqn commented on pkg/agent/nodeportlocal/k8s/annotations.go:
```diff
@@ -84,6 +85,14 @@ func assignPodAnnotation(pod *corev1.Pod, nodeIP string, containerPort, nodePort
+
+	klog.V(2).Infof("Removing annotation from pod: %v", pod.Name)
+	delete(current, NPLAnnotationStr)
+	pod.Annotations = current
```
Since `pod.Annotations` is not too long, maybe just `delete(pod.Annotations, NPLAnnotationStr) to avoid the extra variable and lines?

shubhamavi replied:
Done, thanks

tnqn commented on pkg/agent/nodeportlocal/npl_agent_init.go:
```diff
@@ -80,14 +80,47 @@ func InitController(kubeClient clientset.Interface, portTable *portcache.PortTab
+			return kubeClient.CoreV1().Services(metav1.NamespaceAll).Watch(context.TODO(), metav1.ListOptions{})
+		},
+	}
+	svcCacheStore, svcController := cache.NewInformer(svclw, &corev1.Service{}, resyncPeriod,
```
We should use the `informerFactory` to create a serviceInformer as there's another feature in agent also watching services, otherwise agent would end up with doubling pressure to K8s API and store duplicate resources in memory.

shubhamavi replied:
would you recommend doing this for watching pods as well, in NPL? cc: @monotosh-avi 
Using informerFactory helps addressing another comment from you
> This won't scall well. We should use the PodLister of PodInformer to avoid more queries.

I'm not sure how to access the informer from the current initialisation. Might have to initialise PodInformer separately to avoid kubeclient calls for Pods.

antoninbas replied:
I think I did leave a comment on the initial NPL PR recommending using the shared informer factory to build the informer. However, when it was just Pods that were being watched it wasn't such a big deal as NPL is the only part of the Antrea Agent that watches Pods. I do feel like we should just use it for both Pods & Services now.

tnqn replied:
Originally I thought the Informer doesn't support specifying filter. I just took another look and found it's doable.  
[The informerFactory interface can only set filter for all listers](https://github.com/kubernetes/kubernetes/blob/4b2a98442ea1ffbed9f4335db2a949c3d6a7995e/staging/src/k8s.io/client-go/informers/factory.go#L100-L102), but we could use `NewFilteredPodInformer` to get separate Informer/Indexer for Pod resource, as NPL is still the only one that watches Pods, I think it doesn't matter if it does't use the shared factory to create it. We can revise it if we have another feature that watches Pods in future.

tnqn commented on pkg/agent/nodeportlocal/k8s/controller.go:
```diff
@@ -167,6 +279,12 @@ func (c *Controller) handleDeletePod(key string) error {
+	if obj != nil {
+		newPod := obj.(*corev1.Pod).DeepCopy()
+		removePodAnnotation(newPod)
+		return c.updatePodAnnotation(newPod)
```
At this time, the Pod is already removed from K8s API, I think it will only see NotFound error?

shubhamavi replied:
so in `syncPod` I added an extra conditional for a  podkey which checks if it satisfies the `isNPLEnabledForServiceOfPod` condition.
The podKeys that do not satisfy that condition (meaning that the parent service no longer has the correct annotation), they would land up in handleRemovePod too, in which case the expectation is that the pod annotation should be removed. 
So essentially `obj == nil` means it's deleted Pod case, and `obj != nil` means that the service does not have the annotation (anymore).

tnqn commented on pkg/agent/nodeportlocal/k8s/controller.go:
```diff
@@ -112,6 +122,107 @@ func (c *Controller) EnqueueObj(obj interface{}) {
+	// In case where the app selector in Service gets updated from one valid selector to another
+	// both set of pods (corresponding to old and new selector) need to be considered
+	newSvc, isSvc := newObj.(*corev1.Service)
+	if !isSvc {
```
Could this branch happen? I think only delete event can possiblely get a `cache.DeletedFinalStateUnknown`, not possible for add and update event.

shubhamavi replied:
Removed the check

tnqn commented on pkg/agent/nodeportlocal/k8s/controller.go:
```diff
@@ -112,6 +122,107 @@ func (c *Controller) EnqueueObj(obj interface{}) {
+		return
+	}
+
+	podKeys := append(c.getPodsFromService(oldSvc), c.getPodsFromService(newSvc)...)
```
Do we need to process Pods in both set or just the removed Pods and the added Pods?

shubhamavi replied:
considering a case where the service annotation is updated in an existing Service like so
`nodeportlocal.antrea.io/enabled: "true"` -> `nodeportlocal.antrea.io/enabled: "false"`
getPodsFromService() function won't return any pods for the "new" service because of the annotation check, but we would still want to process the pods corresponding to the old service, which had the proper annotation previously.

shubhamavi replied:
To address this I added a few checks before pushing into the queue, which should also ensure that the two sets are always mutually exclusive, so there should not be a need to check for uniqueness among the two.

tnqn commented on pkg/agent/nodeportlocal/k8s/controller.go:
```diff
@@ -112,6 +122,107 @@ func (c *Controller) EnqueueObj(obj interface{}) {
+func (c *Controller) getPodsFromService(svc *corev1.Service) []string {
+	var pods []string
+	annotations := svc.GetAnnotations()
+	if val, ok := annotations[NPLServiceAnnotation]; !ok || (ok && val != "true") {
```
```suggestion
	if val, ok := annotations[NPLServiceAnnotation]; !ok || val != "true" {
```
because `ok` must be true if the second expression is being evaluated.

shubhamavi replied:
So **labels** and label based filtering is case sensitive in k8s. Therefore, I feel it’s a good idea to make the **annotation** key/value follow similar case sensitivity as well.
In order to verify case sensitivity in labels, we can check via these commands, which give different outputs (if for instance the deployment has `foo=bar` label set)
`kubectl get deployment --selector "foo=bar"`
`kubectl get deployment --selector "foo=BAR"`
I could not find any specific examples in the antrea code, for annotation key/val case sensitivity.

jianjuns replied:
You are right, seems even annotations are typically handled as case sensitive.

tnqn commented on pkg/agent/nodeportlocal/k8s/controller.go:
```diff
@@ -112,6 +122,107 @@ func (c *Controller) EnqueueObj(obj interface{}) {
+		return pods
+	}
+
+	appSelectorValue, ok := svc.Spec.Selector["app"]
```
Is it right? We are not matching Pods in the standard way but requiring the service must have "app" key as a selector and only checking it but not other expressions?

shubhamavi replied:
Could you expand on this a bit, I might be missing something. Isn't defining the pod selector in a service using this scheme the standard way? It assumes  that the service has
```
spec:
  selector:
    app: MyApp
```
and the pods would have 
```
labels:
  app: MyApp
```

I'm essentially following this
https://kubernetes.io/docs/concepts/services-networking/service/ 
> This specification creates a new Service object named "my-service", which targets TCP port 9376 on any Pod with the app=MyApp label.

antoninbas replied:
what Quan means is that there is nothing special about the `app` label. It's just an example in the K8s documentation. I can define Services using a completely different set of labels. For example I can use `foo: bar` as my label selector.

antoninbas replied:
so you need to consider the entire selector

tnqn commented on pkg/agent/nodeportlocal/k8s/controller.go:
```diff
@@ -112,6 +122,107 @@ func (c *Controller) EnqueueObj(obj interface{}) {
+		LabelSelector: labels.SelectorFromSet(labels.Set{"app": appSelectorValue}).String(),
+		FieldSelector: fields.OneTermEqualSelector("spec.nodeName", c.NodeName).String(),
+	}
+	podList, err := c.kubeClient.CoreV1().Pods(metav1.NamespaceAll).List(context.TODO(), listOptions)
```
This won't scall well. We should use the PodLister of PodInformer to avoid more queries.

tnqn commented on pkg/agent/nodeportlocal/k8s/controller.go:
```diff
@@ -112,6 +122,107 @@ func (c *Controller) EnqueueObj(obj interface{}) {
+
+func (c *Controller) isNPLEnabledForServiceOfPod(obj interface{}) bool {
+	pod := obj.(*corev1.Pod)
+	appLabelValue, ok := pod.GetLabels()["app"]
```
ditto

antoninbas commented on pkg/agent/nodeportlocal/k8s/annotations.go:
```diff
@@ -28,6 +28,7 @@ import (
 )
 
 const NPLAnnotationStr = "nodeportlocal.antrea.io"
+const NPLServiceAnnotation = "nodeportlocal.antrea.io/enabled"
```
can you use a `const` block?

```go
const (
    NPLAnnotationStr = "nodeportlocal.antrea.io"
    NPLServiceAnnotation = "nodeportlocal.antrea.io/enabled"
)
```

I would also like the name to be more uniform for both constants. Maybe `NPLAnnotationKey` and `NPLEnabledAnnotationKey`, or similar names.

shubhamavi replied:
Done

antoninbas commented on pkg/agent/nodeportlocal/k8s/annotations.go:
```diff
@@ -84,6 +85,14 @@ func assignPodAnnotation(pod *corev1.Pod, nodeIP string, containerPort, nodePort
+func removePodAnnotation(pod *corev1.Pod) {
+	current := pod.Annotations
+
+	klog.V(2).Infof("Removing annotation from pod: %v", pod.Name)
```
all over the project, we capitalize K8s terms (Pod, Node, Service, ...) in comments / log messages / documentations. Please do the same for consistency.

shubhamavi replied:
Done, thanks

antoninbas commented on pkg/agent/nodeportlocal/k8s/controller.go:
```diff
@@ -69,19 +76,22 @@ func (c *Controller) Run(stopCh <-chan struct{}) {
+
+	go c.PodController.Run(stopCh)
+	go c.SvcController.Run(stopCh)
+	cache.WaitForCacheSync(stopCh, c.PodController.HasSynced)
+	cache.WaitForCacheSync(stopCh, c.SvcController.HasSynced)
```
please use `WaitForNamedCacheSync`

shubhamavi replied:
Done, I have specified the controllerName to be `"AntreaAgentNPLController"`

antoninbas reviewed (commented):
took a quick look, not much to add compared to Quan's very comprehensive review

* the biggest issue IMO is the fact that the PR assumes that Services always use the `app` label exclusively to select endpoints, which is wrong
* we need to add a unit test for the case where a Service selector is updated

jianjuns commented on pkg/agent/nodeportlocal/k8s/controller.go:
```diff
@@ -112,6 +122,107 @@ func (c *Controller) EnqueueObj(obj interface{}) {
+		}
+	}
+
+	for _, podKey := range c.getPodsFromService(svc) {
```
We need to handle Service selector and annotation changes too?

jianjuns replied:
Also do we need synchronization with Pod events for race conditions?
@tnqn

jianjuns replied:
I saw we already check Service changes, so resolve this one.

jianjuns commented on pkg/agent/nodeportlocal/k8s/controller.go:
```diff
@@ -112,6 +122,107 @@ func (c *Controller) EnqueueObj(obj interface{}) {
+		return false
+	}
+
+	for _, service := range c.SvcCacheStore.List() {
```
I am worried about the efficiency of this implementation. Any way to optimize?
How about adding a map for Services with the NPL annotation, if we assume only a small subset of Services have NPL enabled?
@tnqn : any good idea?

shubhamavi replied:
Great suggestion! I have addressed this. Thanks

shubhamavi commented:
Addressed most of the review comments
Another key update as part of the previous commit is to avoid exporting Pod/Service informers from the NPLController. All the eventHandler functions (`enqueuePod/enqueueSvc`) have also been made local to the controller and are not being exported anymore. 
This involved moving the `AddEventHandlerWithResyncPeriod` and `AddIndexers` functions inside `NewNPLController(...)` as well, from `pkg/agent/nodeportlocal/npl_agent_init.go` to `pkg/agent/nodeportlocal/k8s/controller.go`
This approach seems to be in line with how other controllers are initialised as well.

tnqn commented on pkg/agent/nodeportlocal/k8s/controller.go:
```diff
@@ -112,12 +167,109 @@ func (c *Controller) EnqueueObj(obj interface{}) {
+
+// matchSvcSelectorPodLabels verifies that all key/value pairs present in Service's selector
+// are also present in Pod's labels
+func matchSvcSelectorPodLabels(svcSelector, podLabel map[string]string) bool {
```
If selector has no selector, it is assumed to have an external process managing its endpoints, and should not select any Pod. 
See https://kubernetes.io/docs/concepts/services-networking/service/#without-selectors

tnqn commented on pkg/agent/nodeportlocal/npl_agent_init.go:
```diff
@@ -58,36 +56,30 @@ func InitializeNPLAgent(kubeClient clientset.Interface, portRange, nodeName stri
-	c.CacheStore = cacheStore
-	c.Ctrl = controller
+
+	informerFactory := informers.NewSharedInformerFactory(kubeClient, resyncPeriod)
```
This should use the global sharedInformerFactory initialized in cmd package, otherwise there will stil be duplicate connections and storages for service

tnqn commented on pkg/agent/nodeportlocal/k8s/controller.go:
```diff
@@ -64,28 +117,31 @@ func podKeyFunc(pod *corev1.Pod) string {
-	cache.WaitForCacheSync(stopCh, c.Ctrl.HasSynced)
+
+	go c.podInformer.Run(stopCh)
+	go c.svcInformer.Informer().Run(stopCh)
```
I think `informerFactory.Start(stopCh)` will start this. There may be issue to run it twice.

tnqn commented on pkg/agent/nodeportlocal/k8s/controller.go:
```diff
@@ -112,12 +168,115 @@ func (c *Controller) EnqueueObj(obj interface{}) {
+	}
+
+	// aggregating unique podKeys
+	podKeys := sets.NewString(c.getPodsFromService(oldSvc)...).Union(sets.NewString(c.getPodsFromService(newSvc)...))
```
I might be wrong from the begining. To confirm, if a Pod is selected by old service and new service, does it need to be enqueued/processed? I think no as there's actually no change made to it in this case?
If I understand it correctly, it should calculate the different of the two set, instead of the union?

tnqn commented on pkg/agent/nodeportlocal/k8s/controller.go:
```diff
@@ -112,12 +168,115 @@ func (c *Controller) EnqueueObj(obj interface{}) {
+		return pods
+	}
+
+	podList, err := c.podLister.List(labels.SelectorFromSet(labels.Set(svc.Spec.Selector)))
```
Same as `isNPLEnabledForServiceOfPod`, empty selector in service selects nothing, instead of everything.

tnqn commented on pkg/agent/nodeportlocal/k8s/controller.go:
```diff
@@ -112,12 +168,115 @@ func (c *Controller) EnqueueObj(obj interface{}) {
+
+func (c *NPLController) isNPLEnabledForServiceOfPod(obj interface{}) bool {
+	pod := obj.(*corev1.Pod)
+	services, err := c.svcInformer.Informer().GetIndexer().ByIndex(NPLEnabledAnnotationIndex, "true")
```
Better to store `c.svcInformer.Informer()` as a struct member as it calls a factory method everytime it's called while `isNPLEnabledForServiceOfPod` will be called quite often.

tnqn commented on pkg/agent/nodeportlocal/k8s/controller.go:
```diff
@@ -64,28 +117,31 @@ func podKeyFunc(pod *corev1.Pod) string {
+	go c.podInformer.Run(stopCh)
+	go c.svcInformer.Informer().Run(stopCh)
+	cache.WaitForNamedCacheSync(controllerName, stopCh, c.podInformer.HasSynced)
+	cache.WaitForNamedCacheSync(controllerName, stopCh, c.svcInformer.Informer().HasSynced)
```
Not introduced by this PR, but normally it should wait for cachesync first then start workers, otherwise the worker may handle events without having all data synced once, leading to incomplete data, for example, listing pods that selected by a service will be incomplete and waste more CPU cycles to converge.

And waitForNamedCacheSync can take more than one cache, it should exit if it returns false which indicates the data is never synced before stopCh is closed.
```
klog.Infof("Starting %s", statusControllerName)
defer klog.Infof("Shutting down %s", statusControllerName)

if !cache.WaitForNamedCacheSync(statusControllerName, stopCh, c.podInformer.HasSynced, c.svcInformer.Informer().HasSynced) {
    return
}
for i := 0; i < numWorkers; i++ {
    go wait.Until(c.Worker, time.Second, stopCh)
}
```

tnqn reviewed (approved):
LGTM

jianjuns commented on pkg/agent/nodeportlocal/k8s/annotations.go:
```diff
@@ -80,18 +84,23 @@ func assignPodAnnotation(pod *corev1.Pod, nodeIP string, containerPort, nodePort
 	pod.Annotations = current
 }
 
+func removePodAnnotation(pod *corev1.Pod) {
```
Nit: I saw this func is called only once. Probably no need to add a func with two lines.

shubhamavi replied:
The idea was to keep all annotation editing functions under `annotations.go` itself.

jianjuns replied:
Ok. Works for me.

jianjuns commented on pkg/agent/nodeportlocal/npl_agent_init.go:
```diff
@@ -55,39 +53,32 @@ func InitializeNPLAgent(kubeClient clientset.Interface, portRange, nodeName stri
+
+	svcInformer := informerFactory.Core().V1().Services().Informer()
+
+	c := k8s.NewNPLController(kubeClient,
```
Nit: easier to understand if you add an alias for "k8s" like "nplk8s".

jianjuns commented on pkg/agent/nodeportlocal/npl_agent_init.go:
```diff
@@ -55,39 +53,32 @@ func InitializeNPLAgent(kubeClient clientset.Interface, portRange, nodeName stri
+		svcInformer,
+		resyncPeriod,
+		portTable)
+	c.RemoveNPLAnnotationFromPods()
```
Sorry, but why we call this func here?

monotosh-avi replied:
During restart we are removing all IPTABLES rules and annotations, so that we can populate all rules from informer  add/update events and there is no stale entry, in case a POD is deleted. 
This has been improved in the PR for restartability, where we selectively add/delete annotations, instead of deleting everything.

jianjuns replied:
And maybe add some comments to explain it and saying it is a workaround.

shubhamavi replied:
This is supposed to go away as part of https://github.com/vmware-tanzu/antrea/pull/1758, which makes improvements in this logic.

jianjuns commented on pkg/agent/nodeportlocal/k8s/controller.go:
```diff
@@ -112,12 +173,122 @@ func (c *Controller) EnqueueObj(obj interface{}) {
+
+func (c *NPLController) enqueueSvcUpdate(oldObj, newObj interface{}) {
+	// In case where the app selector in Service gets updated from one valid selector to another
+	// both sets of Pods (corresponding to old and new selector) need to be considered
```
Nit: add "." to end of the sentences.

jianjuns commented on pkg/agent/nodeportlocal/k8s/controller.go:
```diff
@@ -112,12 +173,122 @@ func (c *Controller) EnqueueObj(obj interface{}) {
+	// both sets of Pods (corresponding to old and new selector) need to be considered
+
+	// Donot push to queue if 1) Service ResourceVersions don't change OR
+	// 2) Service spec selectors AND 3) NPLEnabledAnnotationKey values don't change
```
Probably we should remove "3)", as the two conditions together define "2)"?

shubhamavi replied:
Updates to .spec.selector and .metadata.annotations can happen separately.

jianjuns commented on pkg/agent/nodeportlocal/k8s/controller.go:
```diff
@@ -112,12 +173,122 @@ func (c *Controller) EnqueueObj(obj interface{}) {
+	}
+
+	// disjunctive union of Pods from both Service sets
+	oldPodSet := sets.NewString(c.getPodsFromService(oldSvc)...)
```
As an optimization, can we get both old and new sets only when selectors changed?

Like:
```
if oldSvc.ResourceVersion == newSvc.ResourceVersion {
    return
}
selectorsChanged := !reflect.DeepEqual(oldSvc.Spec.Selector, newSvc.Spec.Selector)
if !selectorsChanged && oldSvc.Annotations[NPLEnabledAnnotationKey] == newSvc.Annotations[NPLEnabledAnnotationKey] {
    return
}
newPodSet := sets.NewString(c.getPodsFromService(newSvc)...)
if selectorsChanged {
    ...
}
```

shubhamavi replied:
This is covered in the conditional (the negation of it), if the selectors are same we return.

jianjuns replied:
I think if the selectors do not change, but NPL annotation changes, we will still get here? I meant in this case no need to compute both old and new Pod sets, since they are equal.

jianjuns replied:
Actually, probably we better to handle annotation change and selector change separately. If annotation changes, we need not to check selectors at all, but should just either remove old Pod sets or add new Pod sets.

shubhamavi replied:
>no need to compute both old and new Pod sets, since they are equal.

So if the annotation changes from valid (oldSvc) to invalid (newSvc), then the newSvc will map to no pods since it does not have the valid NPL annotation in it (see getPodsFromService). So we'll end up queueing only the pods corresponding to the service with the valid annotation, i.e. oldSvc.

>probably we better to handle annotation change and selector change separately

I didn't quite get the second comment. The code snippet provided by you is pretty much what I have implemented, just squashed into a single conditional block. Are you suggesting to break down the conditional written above into separate blocks, with each block returning if no change is detected? Something like this:

```
if oldSvc.ResourceVersion == newSvc.ResourceVersion {
    return
}

if oldSvc.Annotations[NPLEnabledAnnotationKey] != newSvc.Annotations[NPLEnabledAnnotationKey] {
    newPodSet := sets.NewString(c.getPodsFromService(newSvc)...)
    podKeys := oldPodSet.Difference(newPodSet).Union(newPodSet.Difference(oldPodSet))
    for podKey := range podKeys {
        c.queue.Add(podKey)
    }
    return
}

if !reflect.DeepEqual(oldSvc.Spec.Selector, newSvc.Spec.Selector) {
    newPodSet := sets.NewString(c.getPodsFromService(newSvc)...)
    podKeys := oldPodSet.Difference(newPodSet).Union(newPodSet.Difference(oldPodSet))
    for podKey := range podKeys {
        c.queue.Add(podKey)
    }
    return
}
```

jianjuns replied:
Ok, I got what you meant now. Then if we want to optimize (like the way in the above code block), we can just save the annotation lookup in getPodsFromService(), creating empty Pod sets, Set.Difference() and Set.Union(). I can agree these are not of big overhead.
I always like more (habit from my datapath experience), but I understand you might like to keep code simpler. So, I would let you decide.

jianjuns commented on pkg/agent/nodeportlocal/k8s/controller.go:
```diff
@@ -112,12 +173,122 @@ func (c *Controller) EnqueueObj(obj interface{}) {
+	// disjunctive union of Pods from both Service sets
+	oldPodSet := sets.NewString(c.getPodsFromService(oldSvc)...)
+	newPodSet := sets.NewString(c.getPodsFromService(newSvc)...)
+	podKeys := oldPodSet.Difference(newPodSet).Union(newPodSet.Difference(oldPodSet))
```
Does it mean we do not enqueue the common set of old and new?

shubhamavi replied:
yes, if a pod is present in both the oldSvc and the newSvc sets, then we don't need to enqueue that.
Consider the sets of Service - {Pods}
s1 - {1,2,3} and s2 - {3,4,5}
then we only need to queue {1,2,4,5}, since queueing 3 won't change anything on pod 3.

jianjuns replied:
How if Service's NPL annotation changes?

shubhamavi replied:
same as replied above, if NPL annotation change from valid to invalid, the valid one will lead to a set of pods, while the invalid one won't, because of the `getPodsFromService` function. The function only returns a slice of pods only if the service has the valid NPL annotation.

jianjuns replied:
Got it now. Thanks for the explanation.

jianjuns commented on pkg/agent/nodeportlocal/k8s/controller.go:
```diff
@@ -19,43 +19,95 @@ package k8s
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    c.enqueuePod,
+			DeleteFunc: c.enqueuePod,
+			UpdateFunc: func(old, cur interface{}) {
```
Does it make sense to check if Pod labels changed or not? @tnqn

shubhamavi replied:
Changes to `ContainerPort` will also need to be monitored, as this eventually determines the iptable rule. Same goes for `podIP` as well but i'm not sure in what scenarios would the `podIP` change.
Also changing a `namespace` will affect the cache as the key would change.

We also put an annotation (`nodeportlocal.antrea.io: {nodeIP, nodePort, podPort}`) on the pod after successfully programming the iptable rule, so if someone changes/removes that, and we are not watching for annotation updates, then we won't be able to fix/add back our annotation.

I guess we can keep this one as it is for now.

jianjuns replied:
Sure. We can keep the current version for now. But note none of containerPort, podIP, annotation changes are common, and the Pod event processing is not simple, if we see multiple Pod events are typically generated and cause unnecessary computation, probably we should still consider optimization.

Here can you add some comments to explain what Pod changes (containerPort, IP, labels, NPL annotation) we are interested in? Namespace cannot be changed after a resource is created.

tnqn replied:
> Changes to ContainerPort will also need to be monitored, as this eventually determines the iptable rule. Same goes for podIP as well but i'm not sure in what scenarios would the podIP change.
Also changing a namespace will affect the cache as the key would change.

Most of fields in spec are actually immutable. 
```
spec: Forbidden: pod updates may not change fields other than `spec.containers[*].image`, `spec.initContainers[*].image`, `spec.activeDeadlineSeconds` or `spec.tolerations`
```
So ContainerPort cannot change. `PodIP` may change if Sandbox container is dead as kubelet will recreate it and call CNI again.

jianjuns reviewed (commented):
Sorry for giving comments late.

jianjuns commented on pkg/agent/nodeportlocal/k8s/controller.go:
```diff
@@ -112,12 +173,122 @@ func (c *Controller) EnqueueObj(obj interface{}) {
+	oldSvc := oldObj.(*corev1.Service)
+	if oldSvc.ResourceVersion == newSvc.ResourceVersion ||
+		reflect.DeepEqual(oldSvc.Spec.Selector, newSvc.Spec.Selector) &&
+			oldSvc.Annotations[NPLEnabledAnnotationKey] == newSvc.Annotations[NPLEnabledAnnotationKey] {
```
Another optimization for you to consider: if we assume many services will not have the annotation, we can filter such services earlier. That is neither of oldSvc.Annotations[NPLEnabledAnnotationKey] and newSvc.Annotations[NPLEnabledAnnotationKey] are true, we can just return, and skip reflect.DeepEqual and other code.

jianjuns reviewed (approved):
Thanks for addressing my comments!

antoninbas commented on pkg/agent/nodeportlocal/k8s/controller.go:
```diff
@@ -19,43 +19,97 @@ package k8s
+				curPod := cur.(*corev1.Pod)
+				// Pod fields to watch for: ContainerPort, HostIP, PodIP,
+				// Labels, Annotations (NPLAnnotationKey).
+				if oldPod.ResourceVersion != curPod.ResourceVersion {
+					c.enqueuePod(cur)
+				}
```
If you want to ignore calls to `UpdateFunc` when no change is actually observed, why not also set the resyncPeriod to 0?

shubhamavi replied:
Done thanks

antoninbas commented on pkg/agent/nodeportlocal/k8s/controller.go:
```diff
@@ -112,12 +175,136 @@ func (c *Controller) EnqueueObj(obj interface{}) {
+	newSvc := newObj.(*corev1.Service)
+	oldSvc := oldObj.(*corev1.Service)
+
+	// Return if Service ResourceVersions don't change.
+	if oldSvc.ResourceVersion == newSvc.ResourceVersion {
+		return
+	}
```
why the inconsistency between Pods and Services? For Pods, you perform that check before calling the enqueue function

shubhamavi replied:
This is because there is no separate enqueue function for Pod updates as compared to Service updates (enqueueSvcUpdate). There is a separate function for Service updates because we queue pods from both old and new services. For Pod updates we eventually just enqueue the new Pods.

antoninbas replied:
thanks for the explanation

antoninbas commented on pkg/agent/nodeportlocal/k8s/controller.go:
```diff
@@ -112,12 +175,136 @@ func (c *Controller) EnqueueObj(obj interface{}) {
+func (c *NPLController) checkDeletedSvc(obj interface{}) (*corev1.Service, error) {
+	deletedState, ok := obj.(cache.DeletedFinalStateUnknown)
+	if !ok {
+		return nil, fmt.Errorf("Received unexpected object: %v", obj)
```
we do not capitalize error strings in Antrea, only error logs

antoninbas commented on pkg/agent/nodeportlocal/k8s/controller.go:
```diff
@@ -112,12 +175,136 @@ func (c *Controller) EnqueueObj(obj interface{}) {
+	pod := obj.(*corev1.Pod)
+	services, err := c.svcInformer.GetIndexer().ByIndex(NPLEnabledAnnotationIndex, "true")
+	if err != nil {
+		klog.Errorf("Got error while listing Services with annotation: %v", err)
```
maybe indicate which annotation in the log message, or that could be confusing when inspecting logs

antoninbas commented on pkg/agent/nodeportlocal/nplagent_test.go:
```diff
@@ -94,43 +119,239 @@ func TestMain(t *testing.T) {
+	return data, err
+}
+
+// Add a service (proper annotation) and pod pair, update service with bad
```
we capitalize K8s resources: Pod, Service, Node, etc

(several occurrences in this file)

antoninbas reviewed (commented):
I don't know if this has been mentioned before, but I think the implementation can be improved in the future by leveraging the EndpointSlice API. EndpointSlice resources include a reference to the Service as well as a list of Pods for that Service.

shubhamavi commented:
@jianjuns @antoninbas took care of your latest comments

antoninbas commented on pkg/agent/nodeportlocal/k8s/controller.go:
```diff
@@ -19,43 +19,97 @@ package k8s
+				curPod := cur.(*corev1.Pod)
+				// Pod fields to watch for: ContainerPort, HostIP, PodIP,
+				// Labels, Annotations (NPLAnnotationKey).
+				if oldPod.ResourceVersion != curPod.ResourceVersion {
```
@tnqn do you know if this check is still useful when setting resyncPeriod to 0?

shubhamavi replied:
Might just be a unit test issue, but after removing this I was getting an update storm in the unit tests.

tnqn replied:
It shouldn't receive any dummy update events if resync is disabled. Could you point which unit test receive an update storm? I think there might be something wrong that we should address to avoid unnecessary calls to `UpdateFunc` in case it really happens when actually running.

shubhamavi replied:
yes, looks like the Pod updates (for removing annotations) was being called, while updating the NPL annotation in Service updates. Although I could reproduce this only in the UTs, tried the same set of steps on my kubernetes cluster and everything seemed to be working fine. 
I've fixed it by avoiding irrelevant Pod updates from the NPLController.

antoninbas commented on pkg/agent/nodeportlocal/npl_agent_init.go:
```diff
@@ -17,31 +17,29 @@
 )
 
-const resyncPeriod = 60 * time.Minute
+const resyncPeriod = 0 * time.Minute
```
I think it would be good to explain why we choose 0 with a comment for folks not familiar with `resyncPeriod`. Maybe something like:
```go
// disable periodic calls to the UpdateFunc handler for existing objects. UpdateFunc will only be called when an object is actually updated.
```

antoninbas reviewed (dismissed):
LGTM, I think @jianjuns already approved in an earlier review

antoninbas commented on pkg/agent/nodeportlocal/nplagent_test.go:
```diff
@@ -94,47 +119,244 @@ func TestMain(t *testing.T) {
+
+	// Update Service with bad annotation value.
+	testSvc.Annotations = map[string]string{k8s.NPLEnabledAnnotationKey: "false"}
+	testSvc.ResourceVersion = "2"
```
is explicitly specifying the `ResourceVersion` in tests still needed with the latest version of the controller code?

shubhamavi replied:
correct, Done

antoninbas commented on pkg/agent/nodeportlocal/k8s/annotations.go:
```diff
@@ -80,18 +84,23 @@ func assignPodAnnotation(pod *corev1.Pod, nodeIP string, containerPort, nodePort
 }
 
+func removePodAnnotation(pod *corev1.Pod) {
+	klog.V(2).Infof("Removing all NodePortLocal annotation from Pod: %s/%s", pod.Namespace, pod.Name)
```
nit: why "all"? there is a single annotation

shubhamavi replied:
The logic to remove the annotation was being used at a different place, so I just moved the log from there, and called this function. I believe "all" was used originally since the NPL annotation value in the Pod can consist of multiple nodePort-podPort (`[]NPLAnnotation`) mappings based on the containerPorts mentioned in the Pod, and there is a workflow that has the ability to remove specific entries from that value (`removeFromPodAnnotation`).

antoninbas replied:
maybe it should be `Removing entire NodePortLocal annotation` then? Since "all" implies multiple annotations

antoninbas commented on pkg/agent/nodeportlocal/k8s/annotations.go:
```diff
@@ -122,22 +131,16 @@ func (c *Controller) RemoveNPLAnnotationFromPods() {
+		removePodAnnotation(&podList.Items[i])
 		c.updatePodAnnotation(&podList.Items[i])
 	}
 	klog.Infof("Removed all NodePortLocal annotations from all Pods")
```
same nit as above

antoninbas reviewed (approved):
LGTM, but I want to make sure @tnqn doesn't have any further comments before we merge

tnqn commented on pkg/agent/nodeportlocal/k8s/controller.go:
```diff
@@ -167,11 +342,19 @@ func (c *Controller) handleDeletePod(key string) error {
+	if obj != nil {
+		newPod := obj.(*corev1.Pod).DeepCopy()
+		if _, exists := newPod.Annotations[NPLAnnotationKey]; exists {
+			removePodAnnotation(newPod)
```
Could it make DeepCopy only when the annotation exists?

tnqn reviewed (commented):
LGTM, one comment left

tnqn reviewed (approved):
LGTM

tnqn commented:
/test-all

shubhamavi commented:
/test-e2e

shubhamavi commented:
/test-networkpolicy
